### PR TITLE
[temp] temp's async functions return a Promise

### PR DESCRIPTION
**Pull request by aminya**
The async functions return Promises:
https://github.com/bruce/node-temp/blob/6448b77bb187a6bd72cb9e592d49a587fa6c5639/lib/temp.js#L271

------------

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/temp/index.d.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Type definitions for temp 0.8
+// Type definitions for temp 0.9.4
```
Comment (aminya) [inline types/temp/index.d.ts:1]:
```suggestion
// Type definitions for temp 0.9
```

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
The functions have been promisified. The cases that you've pointed out, the promise will actually be undefined (so you should keep them as void-returning). In the case where no callback has been provided, you'll end up with a Promise of the callback parameter's type.

Seems like the right thing is to split these into overloads.

File: types/temp/index.d.ts [hunk 2]
```diff
@@ -28,17 +28,17 @@ declare namespace temp {
 
   function track(value?: boolean): typeof temp;
 
-  function mkdir(affixes?: string | AffixOptions, callback?: (err: any, dirPath: string) => void): void;
+  function mkdir(affixes?: string | AffixOptions, callback?: (err: any, dirPath: string) => void): Promise<void>;
```
Comment (aminya) [inline types/temp/index.d.ts:31]:
```suggestion
  function mkdir(affixes?: string | AffixOptions, callback: (err: any, dirPath: string) => void): void;
  function mkdir(affixes?: string | AffixOptions): Promise<void>;
```

File: types/temp/index.d.ts [hunk 3]
```diff
@@ -28,17 +28,17 @@ declare namespace temp {
...
 
   function mkdirSync(affixes?: string | AffixOptions): string;
 
-  function open(affixes?: string | AffixOptions, callback?: (err: any, result: OpenFile) => void): void;
+  function open(affixes?: string | AffixOptions, callback?: (err: any, result: OpenFile) => void): Promise<void>;
```
Comment (aminya) [inline types/temp/index.d.ts:35]:
```suggestion
  function open(affixes?: string | AffixOptions, callback: (err: any, result: OpenFile) => void): void;
  function open(affixes?: string | AffixOptions): Promise<void>;
```

File: types/temp/index.d.ts [hunk 4]
```diff
@@ -28,17 +28,17 @@ declare namespace temp {
...
 
   function openSync(affixes?: string | AffixOptions): OpenFile;
 
   function path(affixes?: string | AffixOptions, defaultPrefix?: string): string;
 
-  function cleanup(callback?: (err: any, result: Stats) => void): void;
+  function cleanup(callback?: (err: any, result: Stats) => void): Promise<void>;
```
Comment (aminya) [inline types/temp/index.d.ts:41]:
```suggestion
  function cleanup(callback: (err: any, result: Stats) => void): void;
  function cleanup(): Promise<void>;
```

File: types/temp/temp-tests.ts [hunk 5]
```diff
@@ -12,6 +12,19 @@ function testCleanup() {
             files.toPrecision(4);
         }
     });
+
+    try {
+        const result = await temp.cleanup();
+        if (typeof result === "boolean") {
+            const x = result;
```
Comment (aminya) [inline types/temp/temp-tests.ts:19]: This result is never boolean. I am not sure why the tests are using `boolean`

**Comment by aminya** [comment]
@DanielRosenwasser Could you take a look?

**Comment by aminya** [comment]
@DanielRosenwasser Any news?

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
Thanks for the patience!

**Comment by aminya** [comment]
> Thanks for the patience!

Thank you. You're welcome

Ready to merge

**Comment by aminya** [comment]
Ready to merge

**Pull request merged**